File: Objective-C/advanced/APIDemo/APIDemo/AdMobAdPreloadingViewController.m
//
//  Copyright 2026 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "AdMobAdPreloadingViewController.h"

#import <GoogleMobileAds/GoogleMobileAds.h>
#import <GoogleMobileAds/GoogleMobileAds_Beta.h>

#import "PreloadView.h"

// Preload status messages.
static NSString *const kPreloadAvailable = @"Is available.";
static NSString *const kPreloadExhausted = @"Is exhausted.";

// AdMob test ad unit IDs.
static NSString *const kAppOpenAdUnitID = @"ca-app-pub-3940256099942544/5575463023";
static NSString *const kInterstitialAdUnitID = @"ca-app-pub-3940256099942544/4411468910";
static NSString *const kRewardedAdUnitID = @"ca-app-pub-3940256099942544/1712485313";
static NSString *const kRewardedInterstitialAdUnitID = @"ca-app-pub-3940256099942544/6978759866";

@interface AdMobAdPreloadingViewController () <GADPreloadDelegate, GADFullScreenContentDelegate>

@property(nonatomic, strong) PreloadView *appOpenView;
@property(nonatomic, strong) PreloadView *interstitialView;
@property(nonatomic, strong) PreloadView *rewardedView;
@property(nonatomic, strong) PreloadView *rewardedInterstitialView;

@property(weak, nonatomic) IBOutlet UIStackView *preloadContainer;

@end

@implementation AdMobAdPreloadingViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self startGoogleMobileAdsSDK];
  [self configurePreloadViews];
}

- (void)startGoogleMobileAdsSDK {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    // Initialize the Google Mobile Ads SDK.
    __weak __typeof__(self) weakSelf = self;
    [[GADMobileAds sharedInstance]
        startWithCompletionHandler:^(GADInitializationStatus *_Nonnull status) {
          // Start ad preloading.
          __typeof__(self) strongSelf = weakSelf;
          if (!strongSelf) {
            return;
          }
          [strongSelf startGoogleMobileAdsPreload];
        }];
  });
}

- (void)startGoogleMobileAdsPreload {
  GADRequest *request = [GADRequest request];
  GADPreloadConfigurationV2 *interstitialConfig =
      [[GADPreloadConfigurationV2 alloc] initWithAdUnitID:kInterstitialAdUnitID
                                                  request:request];
  [GADInterstitialAdPreloader.sharedInstance
      preloadForPreloadID:kInterstitialAdUnitID
            configuration:interstitialConfig
                 delegate:self];
  GADPreloadConfigurationV2 *appOpenConfig =
      [[GADPreloadConfigurationV2 alloc] initWithAdUnitID:kAppOpenAdUnitID
                                                  request:request];
  [GADAppOpenAdPreloader.sharedInstance
      preloadForPreloadID:kAppOpenAdUnitID
            configuration:appOpenConfig
                 delegate:self];
  GADPreloadConfigurationV2 *rewardedConfig =
      [[GADPreloadConfigurationV2 alloc] initWithAdUnitID:kRewardedAdUnitID
                                                  request:request];
  [GADRewardedAdPreloader.sharedInstance
      preloadForPreloadID:kRewardedAdUnitID
            configuration:rewardedConfig
                 delegate:self];
  GADPreloadConfigurationV2 *rewardedInterstitialConfig =
      [[GADPreloadConfigurationV2 alloc] initWithAdUnitID:kRewardedInterstitialAdUnitID
                                                  request:request];
  [GADRewardedInterstitialAdPreloader.sharedInstance
      preloadForPreloadID:kRewardedInterstitialAdUnitID
            configuration:rewardedInterstitialConfig
                 delegate:self];
}

- (BOOL)isInterstitialAvailable {
  return [GADInterstitialAdPreloader.sharedInstance
      isAdAvailableWithPreloadID:kInterstitialAdUnitID];
}

- (BOOL)isRewardedAvailable {
  return [GADRewardedAdPreloader.sharedInstance
      isAdAvailableWithPreloadID:kRewardedAdUnitID];
}

- (BOOL)isRewardedInterstitialAvailable {
  return [GADRewardedInterstitialAdPreloader.sharedInstance
      isAdAvailableWithPreloadID:kRewardedInterstitialAdUnitID];
}

- (BOOL)isAppOpenAvailable {
  return [GADAppOpenAdPreloader.sharedInstance
      isAdAvailableWithPreloadID:kAppOpenAdUnitID];
}

- (void)showInterstitialAd {
  if (![self isInterstitialAvailable]) {
    [self logAndShowAlert:@"Preloaded interstitial ad is not available."];
    return;
  }

  // Getting the preloaded ad loads another ad in the background.
  GADInterstitialAd *ad =
      [GADInterstitialAdPreloader.sharedInstance adWithPreloadID:kInterstitialAdUnitID];

  // Interact with the ad object as needed.
  NSLog(@"Interstitial ad response info: %@", ad.responseInfo);
  ad.paidEventHandler = ^(GADAdValue *_Nonnull value) {
    NSLog(@"Interstitial ad paid event: %@ %@ ", value.value, value.currencyCode);
  };
  ad.fullScreenContentDelegate = self;
  [ad presentFromRootViewController:self];
}

- (void)showRewardedAd {
  if (![self isRewardedAvailable]) {
    [self logAndShowAlert:@"Preloaded rewarded ad is not available."];
    return;
  }

  // Getting the preloaded ad loads another ad in the background.
  __weak __typeof__(self) weakSelf = self;
  GADRewardedAd *ad =
      [GADRewardedAdPreloader.sharedInstance adWithPreloadID:kRewardedAdUnitID];

  // Interact with the ad object as needed.
  NSLog(@"Rewarded ad response info: %@", ad.responseInfo);
  ad.paidEventHandler = ^(GADAdValue *_Nonnull value) {
    NSLog(@"Rewarded ad paid event: %@ %@ ", value.value, value.currencyCode);
  };
  ad.fullScreenContentDelegate = self;
  [ad presentFromRootViewController:self
           userDidEarnRewardHandler:^{
             __typeof__(self) strongSelf = weakSelf;
             if (!strongSelf) {
               return;
             }
             GADAdReward *reward = ad.adReward;
             NSLog(@"User was rewarded %@ %@", reward.amount, reward.type);
           }];
}

- (void)showRewardedInterstitialAd {
  if (![self isRewardedInterstitialAvailable]) {
    [self logAndShowAlert:@"Preloaded rewarded interstitial ad is not available."];
    return;
  }

  // Polling returns the next available ad and loads another ad in the background.
  __weak __typeof__(self) weakSelf = self;
  GADRewardedInterstitialAd *ad = [GADRewardedInterstitialAdPreloader.sharedInstance
      adWithPreloadID:kRewardedInterstitialAdUnitID];

  // Interact with the ad object as needed.
  NSLog(@"Rewarded interstitial ad response info: %@", ad.responseInfo);
  ad.paidEventHandler = ^(GADAdValue *_Nonnull value) {
    NSLog(@"Rewarded interstitial ad paid event: %@ %@ ", value.value, value.currencyCode);
  };
  ad.fullScreenContentDelegate = self;
  [ad presentFromRootViewController:self
           userDidEarnRewardHandler:^{
             __typeof__(self) strongSelf = weakSelf;
             if (!strongSelf) {
               return;
             }
             GADAdReward *reward = ad.adReward;
             NSLog(@"User was rewarded %@ %@", reward.amount, reward.type);
           }];
}

- (void)showAppOpenAd {
  if (![self isAppOpenAvailable]) {
    [self logAndShowAlert:@"Preloaded app open ad is not available."];
    return;
  }

  // Getting the preloaded loads another ad in the background.
  GADAppOpenAd *ad =
      [GADAppOpenAdPreloader.sharedInstance adWithPreloadID:kAppOpenAdUnitID];

  // Interact with the ad object as needed.
  NSLog(@"App open ad response info: %@", ad.responseInfo);
  ad.paidEventHandler = ^(GADAdValue *_Nonnull value) {
    NSLog(@"App open ad paid event: %@ %@ ", value.value, value.currencyCode);
  };
  ad.fullScreenContentDelegate = self;
  [ad presentFromRootViewController:self];
}

- (void)configurePreloadViews {
  __weak __typeof__(self) weakSelf = self;
  self.interstitialView = [PreloadView preloadViewWithTitle:@"Interstitial"
                                               showDelegate:^{
                                                 [weakSelf showInterstitialAd];
                                               }];
  self.rewardedView = [PreloadView preloadViewWithTitle:@"Rewarded"
                                           showDelegate:^{
                                             [weakSelf showRewardedAd];
                                           }];
  self.rewardedInterstitialView = [PreloadView preloadViewWithTitle:@"Rewarded Interstitial"
                                                       showDelegate:^{
                                                         [weakSelf showRewardedInterstitialAd];
                                                       }];
  self.appOpenView = [PreloadView preloadViewWithTitle:@"App open"
                                          showDelegate:^{
                                            [weakSelf showAppOpenAd];
                                          }];

  [self.preloadContainer addArrangedSubview:self.interstitialView];
  [self.preloadContainer addArrangedSubview:self.appOpenView];
  [self.preloadContainer addArrangedSubview:self.rewardedView];
  [self.preloadContainer addArrangedSubview:self.rewardedInterstitialView];

  [self updatePreloadViews];
}

- (void)updatePreloadViews {
  [self updatePreloadView:self.interstitialView isAvailable:[self isInterstitialAvailable]];
  [self updatePreloadView:self.rewardedView isAvailable:[self isRewardedAvailable]];
  [self updatePreloadView:self.rewardedInterstitialView isAvailable:[self isRewardedInterstitialAvailable]];
  [self updatePreloadView:self.appOpenView isAvailable:[self isAppOpenAvailable]];
}

- (void)updatePreloadView:(PreloadView *)preloadView isAvailable:(BOOL)isAvailable {
  preloadView.showButton.enabled = isAvailable;
  UILabel *statusText = preloadView.statusText;
  statusText.text = isAvailable ? kPreloadAvailable : kPreloadExhausted;
  statusText.textColor = isAvailable ? UIColor.blueColor : UIColor.redColor;
  UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification,
                                  statusText.text);
}

- (void)logAndShowAlert:(NSString *)message {
  UIAlertController *alert =
      [UIAlertController alertControllerWithTitle:@"Alert"
                                          message:message
                                   preferredStyle:UIAlertControllerStyleAlert];

  UIAlertAction *dismissAction = [UIAlertAction actionWithTitle:@"Dismiss"
                                                          style:UIAlertActionStyleDefault
                                                        handler:nil];
  [alert addAction:dismissAction];
  [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - GADPreloadDelegate

- (void)adAvailableForPreloadID:(nonnull NSString *)preloadID responseInfo:(nonnull GADResponseInfo *)responseInfo {
  // This callback indicates that an ad is available for the specified configuration.
  // No action is required here, but updating the UI can be useful in some cases.
  NSLog(@"Ad preloaded successfully for ad unit ID: %@", preloadID);
  [self updatePreloadViews];
}

- (void)adsExhaustedForPreloadID:(nonnull NSString *)preloadID {
  // This callback indicates that all the ads for the specified configuration have been
  // consumed and no ads are available to show. No action is required here, but updating
  // the UI can be useful in some cases.
  // [Important] Don't call [GAD<Format>AdPreloader preloadForPreloadID:] or
  // [GAD<Format>AdPreloader adWithPreloadID:] from adsExhaustedForPreloadID.
  NSLog(@"Ad exhausted for ad preload ID: %@", preloadID);
  [self updatePreloadViews];
}

- (void)adFailedToPreloadForPreloadID:(nonnull NSString *)preloadID error:(nonnull NSError *)error {
  NSLog(@"Ad failed to load with ad preload ID: %@, Error: %@", preloadID, error.localizedDescription);
}

#pragma mark - GADFullScreenContentDelegate

- (void)adWillPresentFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"Preloaded ad will be presented.");
}

- (void)adDidDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"Preloaded ad dismissed.");
}

@end

File: Swift/advanced/APIDemo/APIDemo/AdMobAdPreloadingViewController.swift
//
//  Copyright 2026 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import GoogleMobileAds
import GoogleMobileAds_Private
import UIKit

class AdMobAdPreloadingViewController: UIViewController, PreloadDelegate,
  FullScreenContentDelegate
{

  /// Status messages for preload views.
  enum StatusMessages {
    static let preloadAvailable = "Is available."
    static let preloadExhausted = "Is exhausted."
  }

  /// AdMob ad unit IDs for testing.
  enum AdUnitIDs {
    static let appOpenAdUnitID = "ca-app-pub-3940256099942544/5575463023"
    static let interstitialAdUnitID = "ca-app-pub-3940256099942544/4411468910"
    static let rewardedAdUnitID = "ca-app-pub-3940256099942544/1712485313"
    static let rewardedInterstitialAdUnitID = "ca-app-pub-3940256099942544/6978759866"
  }

  @IBOutlet private weak var preloadContainer: UIStackView!

  private var isMobileAdsStartCalled = false
  private var interstitialView: PreloadView!
  private var rewardedView: PreloadView!
  private var rewardedInterstitialView: PreloadView!
  private var appOpenView: PreloadView!

  override func viewDidLoad() {
    super.viewDidLoad()
    startGoogleMobileAdsSDK()
    configurePreloadViews()
  }

  private func startGoogleMobileAdsSDK() {
    DispatchQueue.main.async {
      guard !self.isMobileAdsStartCalled else { return }

      self.isMobileAdsStartCalled = true

      // Initialize the Google Mobile Ads SDK.
      MobileAds.shared.start { [weak self] status in
        guard let self else { return }
        // Start ad preloading.
        self.startGoogleMobileAdsPreload()
      }
    }
  }

  private func startGoogleMobileAdsPreload() {
    let request = Request()
    let interstitialConfig = PreloadConfigurationV2(
      adUnitID: AdUnitIDs.interstitialAdUnitID, request: request)
    InterstitialAdPreloader.shared.preload(
      for: AdUnitIDs.interstitialAdUnitID, configuration: interstitialConfig, delegate: self)

    let rewardedConfig = PreloadConfigurationV2(
      adUnitID: AdUnitIDs.rewardedAdUnitID, request: request)
    RewardedAdPreloader.shared.preload(
      for: AdUnitIDs.rewardedAdUnitID, configuration: rewardedConfig, delegate: self)

    let appOpenConfig = PreloadConfigurationV2(
      adUnitID: AdUnitIDs.appOpenAdUnitID, request: request)
    AppOpenAdPreloader.shared.preload(
      for: AdUnitIDs.appOpenAdUnitID, configuration: appOpenConfig, delegate: self)

    let rewardedInterstitialConfig = PreloadConfigurationV2(
      adUnitID: AdUnitIDs.rewardedInterstitialAdUnitID, request: request)
    RewardedInterstitialAdPreloader.shared.preload(
      for: AdUnitIDs.rewardedInterstitialAdUnitID, configuration: rewardedInterstitialConfig,
      delegate: self)
  }

  private func isInterstitialAvailable() -> Bool {
    return InterstitialAdPreloader.shared.isAdAvailable(with: AdUnitIDs.interstitialAdUnitID)
  }

  private func isRewardedAvailable() -> Bool {
    return RewardedAdPreloader.shared.isAdAvailable(with: AdUnitIDs.rewardedAdUnitID)
  }

  private func isAppOpenAvailable() -> Bool {
    return AppOpenAdPreloader.shared.isAdAvailable(with: AdUnitIDs.appOpenAdUnitID)
  }

  private func isRewardedInterstitialAvailable() -> Bool {
    return RewardedInterstitialAdPreloader.shared.isAdAvailable(
      with: AdUnitIDs.rewardedInterstitialAdUnitID)
  }

  private func showInterstitialAd() {
    // Polling returns the next available ad and loads another ad in the background.
    guard let ad = InterstitialAdPreloader.shared.ad(with: AdUnitIDs.interstitialAdUnitID) else {
      printAndShowAlert("Preload interstitial ad is exhausted.")
      return
    }

    // Interact with the ad object as needed.
    print("Interstitial ad response info: \(ad.responseInfo)")
    ad.paidEventHandler = { (value: AdValue) in
      print("Interstitial ad paid event: \(value.value), \(value.currencyCode)")
    }

    ad.fullScreenContentDelegate = self
    ad.present(from: self)
  }

  private func showRewardedAd() {
    // Polling returns the next available ad and loads another ad in the background.
    guard let ad = RewardedAdPreloader.shared.ad(with: AdUnitIDs.rewardedAdUnitID) else {
      printAndShowAlert("Preloaded rewarded ad is exhausted.")
      return
    }

    // Interact with the ad object as needed.
    print("Rewarded ad response info: \(ad.responseInfo)")
    ad.paidEventHandler = { (value: AdValue) in
      print("Rewarded ad paid event: \(value.value), \(value.currencyCode)")
    }

    ad.fullScreenContentDelegate = self
    ad.present(from: self) {
      let reward = ad.adReward
      print("User was rewarded \(reward.amount) \(reward.type)")
    }
  }

  private func showAppOpenAd() {
    // Polling returns the next available ad and loads another ad in the background.
    guard let ad = AppOpenAdPreloader.shared.ad(with: AdUnitIDs.appOpenAdUnitID) else {
      printAndShowAlert("Preload app open ad is exhausted.")
      return
    }

    // Interact with the ad object as needed.
    print("App open ad response info: \(ad.responseInfo)")
    ad.paidEventHandler = { (value: AdValue) in
      print("App open ad paid event: \(value.value), \(value.currencyCode)")
    }

    ad.fullScreenContentDelegate = self
    ad.present(from: self)
  }

  private func showRewardedInterstitialAd() {
    // Polling returns the next available ad and loads another ad in the background.
    guard
      let ad = RewardedInterstitialAdPreloader.shared.ad(
        with: AdUnitIDs.rewardedInterstitialAdUnitID)
    else {
      printAndShowAlert("Preload rewarded interstitial ad is exhausted.")
      return
    }

    // Interact with the ad object as needed.
    print("Rewarded interstitial ad response info: \(ad.responseInfo)")
    ad.paidEventHandler = { (value: AdValue) in
      print("Rewarded interstitial ad paid event: \(value.value), \(value.currencyCode)")
    }

    ad.fullScreenContentDelegate = self
    ad.present(from: self) {
      let reward = ad.adReward
      print("User was rewarded \(reward.amount) \(reward.type)")
    }
  }

  private func configurePreloadViews() {
    interstitialView = PreloadView.load(
      title: "Interstitial",
      showDelegate: { [weak self] in
        self?.showInterstitialAd()
      })
    preloadContainer.addArrangedSubview(interstitialView)
    rewardedView = PreloadView.load(
      title: "Rewarded",
      showDelegate: { [weak self] in
        self?.showRewardedAd()
      })
    preloadContainer.addArrangedSubview(rewardedView)
    appOpenView = PreloadView.load(
      title: "App open",
      showDelegate: { [weak self] in
        self?.showAppOpenAd()
      })
    preloadContainer.addArrangedSubview(appOpenView)
    rewardedInterstitialView = PreloadView.load(
      title: "Rewarded Interstitial",
      showDelegate: { [weak self] in
        self?.showRewardedInterstitialAd()
      })
    preloadContainer.addArrangedSubview(rewardedInterstitialView)

    updatePreloadViews()
  }

  private func updatePreloadViews() {
    updatePreloadView(preloadView: interstitialView, isAvailable: isInterstitialAvailable())
    updatePreloadView(preloadView: rewardedView, isAvailable: isRewardedAvailable())
    updatePreloadView(preloadView: appOpenView, isAvailable: isAppOpenAvailable())
    updatePreloadView(
      preloadView: rewardedInterstitialView, isAvailable: isRewardedInterstitialAvailable())
  }

  private func updatePreloadView(preloadView: PreloadView, isAvailable: Bool) {
    preloadView.showButton.isEnabled = isAvailable
    preloadView.statusText.text =
      isAvailable ? StatusMessages.preloadAvailable : StatusMessages.preloadExhausted
    preloadView.statusText.textColor = isAvailable ? UIColor.blue : UIColor.red
    if let statusText = preloadView.statusText.text {
      UIAccessibility.post(notification: .announcement, argument: statusText)
    }
  }

  private func printAndShowAlert(_ message: String) {
    let alert = UIAlertController(title: "Alert", message: message, preferredStyle: .alert)
    let dismissAction = UIAlertAction(title: "Dismiss", style: .default)
    alert.addAction(dismissAction)
    present(alert, animated: true, completion: nil)
  }

  // MARK: - PreloadDelegate
  func adAvailable(forPreloadID preloadID: String, responseInfo: ResponseInfo) {
    // This callback indicates that an ad is available for the specified configuration.
    // No action is required here, but updating the UI can be useful in some cases.
    print("Ad preloaded successfully for ad preload ID: \(preloadID)")
    updatePreloadViews()
  }

  func adsExhausted(forPreloadID preloadID: String) {
    // This callback indicates that all the ads for the specified configuration have been
    // consumed and no ads are available to show. No action is required here, but updating
    // the UI can be useful in some cases.
    // [Important] Don't call AdPreloader.shared.preload or AdPreloader.shared.ad
    // from adsExhausted.
    print("Ad exhausted for ad preload ID: \(preloadID)")
    updatePreloadViews()
  }

  func adFailedToPreload(forPreloadID preloadID: String, error: Error) {
    print(
      "Ad failed to load with ad preload ID: \(preloadID), Error: \(error.localizedDescription)"
    )
  }

  // MARK: - FullScreenContentDelegate

  func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
    print("Preloaded ad will be presented.")
  }

  func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
    print("Preloaded ad dismissed.")
  }
}
